WooCommerce/Classes/POS/Presentation/ItemListView.swift

```diff
@@ -114,48 +133,56 @@ private extension ItemListView {
+                            .transition(.opacity.combined(with: .move(edge: .trailing)))
+                            .onChange(of: searchTerm) { oldValue, newValue in
+                                // The debouncing logic is a little tricky, because the loading state is held in the controller.
+                                // Arguably, we should use view state `isSearching` for this, so the UI is independent of the request timing.
+
+                                // As the user types, we don't want to send every keystroke to the remote, so we debounce the requests.
+                                // However, we don't want to debounce the first keystroke of a new search, so that the loading
+                                // state shows immediately and the UI feels responsive.
+
+                                // So, if the last search was finished, we don't debounce the first character. If it didn't
+                                // finish i.e. it is still ongoing, we debounce the next keystrokes by 300ms. In either case,
+                                // the ongoing search is redundant now there's a new search term, so we cancel it.
+                                let shouldDebounceNextSearchRequest = !didFinishSearch
+                                searchTask?.cancel()
+
+                                searchTask = Task {
+                                    if shouldDebounceNextSearchRequest {
+                                        try? await Task.sleep(nanoseconds: 300 * NSEC_PER_MSEC)
+                                    }
+
+                                    guard !Task.isCancelled else { return }
+
+                                    didFinishSearch = false
+
+                                    await posModel.purchasableItemsSearchController.searchItems(searchTerm: newValue, baseItem: .root)
+
+                                    if !Task.isCancelled {
+                                        didFinishSearch = true
+                                    }
                                 }
                             }
+
+                        POSPageHeaderActionButton(systemName: "magnifyingglass") {
+                            selectedItemListType = .products(search: true)
+                            isSearchFieldFocused = true
                         }
+                        .renderedIf(!shouldShowSearchField)
+                        .transition(.opacity.combined(with: .scale))
                     }
 
                     if shouldShowCoupons {
                         POSPageHeaderActionButton(systemName: "plus") {
                             ServiceLocator.analytics.track(.pointOfSaleCouponsCreateTapped)
                             showCouponCreationModal = true
                         }
-                        .opacity(isAddingCouponAllowed ? 1 : 0)
+                        .renderedIf(isAddingCouponAllowed)
+                        .transition(.opacity.combined(with: .scale))
```

Comment:
I used opacity to "reserve" a place for the button to avoid issues when switching between Products and Coupons with large dynamic type. Those are quite exceptional cases. There won't be issues while the number of buttons matches in Coupons and Products.

As I understand, some of that is solved by limiting the dynamic type size for the header?

Reply:
Either way, I love the animation!

Reply:
Yep, I know, and agree that it's rare I looked at trying to keep the reserved space, but it was over complicated and didn't bring much value. 

It's mainly fixed by limiting the size of the titles.